### ISSUE TITLE
[JENKINS-14618] ignore the checkout on the master

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/matrix-project-plugin#516]
<p>in a Matrix job , we need to ability to ignore the checkout on the Master. </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=morlajb">morlajb</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618" target="_blank">ignore the checkout on the master</a></i></summary>
<i><ul>
<li><b>status</b>: Reopened
<li><b>priority</b>: Critical
<li><b>component(s)</b>: matrix-project-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 15
<li><b>watchers</b>: 19
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>in a Matrix job , we need to ability to ignore the checkout on the Master. </p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>all</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/14618?focusedId=213276&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213276">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618?focusedId=213276&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213276">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Duplicates <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a></p>
</pre>
</details>

</ul>
</details>

<p>Duplicates <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/14618?focusedId=213315&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213315">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618?focusedId=213315&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213315">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Oleg: I disagree, the other wants 1 SCM checkout, this one wants N checkouts, instead of N+1. This one is actually the exact opposite of <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a>. They're related, but not duplicate IMO.</p>
</pre>
</details>

</ul>
</details>

<p>Oleg: I disagree, the other wants 1 SCM checkout, this one wants N checkouts, instead of N+1. This one is actually the exact opposite of <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a>. They're related, but not duplicate IMO.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>isidius:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/14618?focusedId=213336&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213336">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618?focusedId=213336&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213336">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have a similar case:</p>

<p>I have a multi-configuration Job that has to run on multiple nodes at the same time. I want the job to checkout from CVS, then do build steps on the CVS content.
Current situation
-Multiple slave nodes are assigned to the job. 
-Master is NOT assigned to the job.
-CVS is accessed by pserver CVSROOT and defined in the job configuration
Current behaviour
-Master checks out from CVS, then delegates job to the slave nodes, which also do a CVS checkout, then do the job
Expected behaviour
-Master immediately delegates to the assigned nodes without doing a CVS chekout first. Since the master is not an assigned node of the job, it is not supposed to do anything concerning the job.</p>
</pre>
</details>

</ul>
</details>

<p>I have a similar case:</p>

<p>I have a multi-configuration Job that has to run on multiple nodes at the same time. I want the job to checkout from CVS, then do build steps on the CVS content.<br/>
Current situation<br/>
-Multiple slave nodes are assigned to the job. <br/>
-Master is NOT assigned to the job.<br/>
-CVS is accessed by pserver CVSROOT and defined in the job configuration<br/>
Current behaviour<br/>
-Master checks out from CVS, then delegates job to the slave nodes, which also do a CVS checkout, then do the job<br/>
Expected behaviour<br/>
-Master immediately delegates to the assigned nodes without doing a CVS chekout first. Since the master is not an assigned node of the job, it is not supposed to do anything concerning the job.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/14618?focusedId=213338&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213338">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618?focusedId=213338&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213338">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Reverted the status. I just imagine a plugin with several checkboxes for masters and slaves, hence these issues have a similar implementation for me. </p>

<p>You're right that feature requests are different in general</p>
</pre>
</details>

</ul>
</details>

<p>Reverted the status. I just imagine a plugin with several checkboxes for masters and slaves, hence these issues have a similar implementation for me. </p>

<p>You're right that feature requests are different in general</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tommigun.jpg" title="tommigun's avatar" /> tommigun:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/14618?focusedId=254561&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254561">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618?focusedId=254561&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254561">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can this be reopened? There are other issues that had been closed as duplicates of this, and this one is closed. I see it if not a bug, then a huge oversight that there is not a checkbox for whether master should do an SCM checkout as well. I am in the exact same boat as Alexander, the master does a lengthy unnecessary SCM checkout before the slaves start building. I tried to work around this by using a custom workspace for the slaves and reuse it from the master, but the workspace name can not be constructed from the axis values as they are not initialised there yet. The current behaviour is 1) unexpected 2) increases build times by tens of percent for no gain. What good does it do to do two SCM syncs, where one is discarded?</p>
</pre>
</details>

</ul>
</details>

<p>Can this be reopened? There are other issues that had been closed as duplicates of this, and this one is closed. I see it if not a bug, then a huge oversight that there is not a checkbox for whether master should do an SCM checkout as well. I am in the exact same boat as Alexander, the master does a lengthy unnecessary SCM checkout before the slaves start building. I tried to work around this by using a custom workspace for the slaves and reuse it from the master, but the workspace name can not be constructed from the axis values as they are not initialised there yet. The current behaviour is 1) unexpected 2) increases build times by tens of percent for no gain. What good does it do to do two SCM syncs, where one is discarded?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/14618?focusedId=254566&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254566">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618?focusedId=254566&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254566">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>Can this be reopened? There are other issues that had been closed as duplicates of this, and this one is closed.</p></blockquote>

<p>This has been reopened in late 2014 and not closed since.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>Can this be reopened? There are other issues that had been closed as duplicates of this, and this one is closed.</p></blockquote>

<p>This has been reopened in late 2014 and not closed since.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>sjmk:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/14618?focusedId=289686&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-289686">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618?focusedId=289686&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-289686">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can't this be resolved somehow? In combination with this <a href="https://issue-redirect.jenkins.io/issue/22494" title="Multiconfiguration project does not respect label restrictions" class="issue-link" data-issue-key="JENKINS-22494">JENKINS-22494</a> and a lot of slave machines, this is pretty annoying, since a random slave is used and repositories are checked out there (which takes up disk space). Is there a workaround?</p>
</pre>
</details>

</ul>
</details>

<p>Can't this be resolved somehow? In combination with this <a href="https://issue-redirect.jenkins.io/issue/22494" title="Multiconfiguration project does not respect label restrictions" class="issue-link" data-issue-key="JENKINS-22494">JENKINS-22494</a> and a lot of slave machines, this is pretty annoying, since a random slave is used and repositories are checked out there (which takes up disk space). Is there a workaround?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>env:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/14618?focusedId=381535&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381535">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-14618?focusedId=381535&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381535">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there any way to to force fix for this issue?</p>

<p>We have huge repository with multiple branches, this additional checkout significantly increases build time.</p>
</pre>
</details>

</ul>
</details>

<p>Is there any way to to force fix for this issue?</p>

<p>We have huge repository with multiple branches, this additional checkout significantly increases build time.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/matrix-project-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-25040%22">JENKINS-25040</a>]</i>
